### DISCUSSION TITLE
Can i call local function in getStaticProps for dynamic routing?

### DISCUSSION THREAD

**Discussion by MoZZANG** [vercel/next.js#50205] [Help]
### Summary

i want to make dynamic routing for recycling a page. And that page is almost static except for button. so what i thought is dynamic routing with local switching function. But my co-worker said 'getStaticProps function always call not just internal(local)api but external api, so your code can't fit for purpose of getStaticProps func.'  Is it?

I'm looking for the reason and how can i use getStaticProps func right way, tell me what your opinion guys.

![image](https://github.com/vercel/next.js/assets/98200337/ce77111b-85fc-4780-af6c-899b9559d904)
![image](https://github.com/vercel/next.js/assets/98200337/1623f19a-8a4c-4829-98c2-5295992aebef)
![image](https://github.com/vercel/next.js/assets/98200337/ee7035ee-89d8-43c6-95d3-1a6048e70a93)

**Comment by icyJoseph** [comment]
I don't understand... I think your co-worker means, that you cannot really expect to be able to do `fetch("/api/some-api-route")`, inside GetStaticProps, because, at build time, there's no server to handle that request.

However, your example is just calling a function. Which doesn't even need to be awaited, it just a synchronous call. Talking about `returnCompleteCnt`

**Reply by MoZZANG** [reply]
Thank you for replying. Ok, first, it's true that i call just function no need to await.

Second, what my co-worker said was that getStaticProps function is supposed to be for calling external api, not just API routes or a function which bring just static text at somewhere in util folder under src.

So what i wanna know is getStaticProps function always use to call external api(e.g rest api)? That means my code which is there is not good example?

**Reply by icyJoseph** [reply]
The Next.js docs say:

> [When should I use getStaticProps?](https://nextjs.org/docs/pages/building-your-application/data-fetching/get-static-props#when-should-i-use-getstaticprops)
>
> You should use getStaticProps if:
> 
> - The data required to render the page is available at build time ahead of a user’s request
> - The data comes from a headless CMS
> - The page must be pre-rendered (for SEO) and be very fast — getStaticProps generates HTML and JSON files, both of which can be cached by a CDN for performance
> - The data can be publicly cached (not user-specific). This condition can be bypassed in certain specific situation by using a Middleware to rewrite the path.

Perhaps, 99% of cases use `getStaticProps` to fetch data but it is not the only use case. I don't understand the limitation imposed by your co-worker. What if you just wanted to generate some random strings, and use them as props for the page? What if you just want to read some JSON? 

In your case it might even be a good idea, because otherwise, you'd need to call the `returnCompleteCtn` function, every time the page renders. Your approach allows Next.js to inline the data statically, and eliminate `returnCompleteCtn` (potentially) from the client bundle, and with it any function calls that are totally unnecessary for static data.
